gr-blocks/lib/qa_gr_hier_block2_derived.cc

```diff
@@ -72,3 +72,33 @@ BOOST_AUTO_TEST_CASE(test_1)
 
     tb->run();
 }
+
+BOOST_AUTO_TEST_CASE(test_2)
+{
+    gr_derived_block_sptr blk(gr_make_derived_block());
+    blk->lock();
+    blk->unlock();
+}
+
+BOOST_AUTO_TEST_CASE(test_3)
+{
+    gr::top_block_sptr tb(gr::make_top_block("test"));
+
+    gr::block_sptr src(gr::blocks::null_source::make(sizeof(int)));
+    gr::block_sptr head(gr::blocks::head::make(sizeof(int), 1000));
+    gr_derived_block_sptr blk(gr_make_derived_block());
+    gr::block_sptr dst0(gr::blocks::null_sink::make(sizeof(int)));
+    gr::block_sptr dst1(gr::blocks::null_sink::make(sizeof(int)));
+
+    tb->connect(src, 0, head, 0);
+    tb->connect(head, 0, blk, 0);
+    tb->connect(blk, 0, dst0, 0);
+    tb->connect(blk, 0, dst1, 0);
+
+    tb->run();
+
+    tb->disconnect(blk, 0, dst1, 0);
+    blk->lock();
+    blk->unlock();
+    tb->run();
```

Comment:
As with the Python test, I think the order should be changed here:
```suggestion
    blk->lock();
    tb->disconnect(blk, 0, dst1, 0);
    blk->unlock();
    tb->wait();
```
The test still crashes on master & passes on this branch.

Reply:
No.
We should disconnect first to trigger the bug, then call `lock()` to get the freeze happen.

Reply:
I guess it depends whether you want to test `lock()` or `unlock()`. I believe they both had the same bug, so it might actually be good to test both.